src/_pytest/cacheprovider.py

```diff
@@ -203,17 +205,35 @@ def set(self, key: str, value: object) -> None:
             with f:
                 f.write(data)
 
-    def _ensure_supporting_files(self) -> None:
-        """Create supporting files in the cache dir that are not really part of the cache."""
-        readme_path = self._cachedir / "README.md"
-        readme_path.write_text(README_CONTENT, encoding="UTF-8")
+    def _ensure_cache_dir_and_supporting_files(self) -> None:
+        """Create the cache dir and its supporting files."""
+        if self._cachedir.is_dir():
+            return
 
-        gitignore_path = self._cachedir.joinpath(".gitignore")
-        msg = "# Created by pytest automatically.\n*\n"
-        gitignore_path.write_text(msg, encoding="UTF-8")
+        files: Iterable[Tuple[str, Union[str, bytes]]] = (
+            ("README.md", README_CONTENT),
+            (".gitignore", "# Created by pytest automatically.\n*\n"),
+            ("CACHEDIR.TAG", CACHEDIR_TAG_CONTENT),
+        )
 
-        cachedir_tag_path = self._cachedir.joinpath("CACHEDIR.TAG")
-        cachedir_tag_path.write_bytes(CACHEDIR_TAG_CONTENT)
+        with tempfile.TemporaryDirectory() as d:
+            for file, content in files:
+                file = os.path.join(d, file)
+                if isinstance(content, str):
+                    with open(file, "xt", encoding="UTF-8") as f:
+                        f.write(content)
+                elif isinstance(content, bytes):
+                    with open(file, "xb") as f:
+                        f.write(content)
+                else:
+                    assert_never(content)
+            shutil.move(d, self._cachedir)
```

Comment:
We should handle the a possible race condition here: we do check if `self._cachedir` does not exist at the start of this method, however some other process might have created the directory since then and this will raise an error I believe.

Reply:
Actually I think we should use `os.rename` explicitly instead of `shutil.move`.

[`os.rename`](https://docs.python.org/3/library/os.html#os.rename):

> On Windows, if dst exists a FileExistsError is always raised. The operation may fail if src and dst are on different filesystems. Use shutil.move() to support moves to a different filesystem.

> On Unix, ... If dst is a non-empty directory, an OSError is raised.

[`shutil.move`](https://docs.python.org/3/library/shutil.html#shutil.move):

> If dst is an existing directory or a symlink to a directory, then src is moved inside that directory. The destination path in that directory must not already exist.

> If dst already exists but is not a directory, it may be overwritten depending on [os.rename()](https://docs.python.org/3/library/os.html#os.rename) semantics.

> If the destination is on the current filesystem, then [os.rename()](https://docs.python.org/3/library/os.html#os.rename) is used. Otherwise, src is copied to the destination using copy_function and then removed. In case of symlinks, a new symlink pointing to the target of src will be created as the destination and src will be removed.

If the directory exists by the time this line is reached, `shutil.move` will move `cachedir` inside the existing directory.

Reply:
I'm using shutil.move because it deals with the other-filesystem case. I can try to be more explicit but it'll amount to a partial reimplementation of shuttle.move.

Reply:
Done.